core/repository/api/src/main/java/org/eclipse/rdf4j/repository/base/AbstractRepositoryConnection.java

```diff
@@ -366,10 +370,23 @@ public void add(Iterable<? extends Statement> statements, Resource... contexts)
 				"contexts argument may not be null; either the value should be cast to Resource or an empty array should be supplied");
 
 		boolean localTransaction = startLocalTransaction();
+		Set<Namespace> newNamespaces = new HashSet<>();
+
 		try {
 			for (Statement st : statements) {
+				if (st instanceof NamespaceAware) {
+					newNamespaces.addAll(((NamespaceAware) st).getNamespaces());
+				}
```

Comment:
I don't think this will work - it's not `Statement` that implements `NamespaceAware`, but the collection of statements.

It's also potentially a bit of a performance burden doing the check this way. I'd suggest that instead, you do a single check outside of the loop, and then just transfer all namespaces in one go, something like this.:

```java
if (statements instanceof NamespaceAware) {
   var namespaces = ((NamespaceAware)statements).getNamespaces();
   for (namespace: namespaces) {
     setNamespace(namespace.getPrefix(), namespace.getName());
   } 
}
```

I also don't think there's need to check if the namespace already exists in the repo or not. `setNamespace` will just overwrite if it exists (which I think is the behavior we want here).

Reply:
I'll get to this today, however, didn't you suggest that we skip existing namespaces in the original issue I created?

Which option should I go with?

Reply:
> I'll get to this today, however, didn't you suggest that we skip existing namespaces in the original issue I created?
> 
> Which option should I go with?

Ah good call I had forgotten that. Yes, let's go with skipping existing for now, as it's more consistent with the behavior when adding a new file.